pkg/apis/extensions/types.go

```diff
@@ -278,6 +283,40 @@ type DeploymentStatus struct {
 
 	// Total number of unavailable pods targeted by this deployment.
 	UnavailableReplicas int32 `json:"unavailableReplicas,omitempty"`
+
+	// Represents the latest available observations of a deployment's current state.
+	Conditions []DeploymentCondition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
+}
+
+type DeploymentConditionType string
+
+// These are valid conditions of a deployment.
+const (
+	// DeploymentAvailable means the deployment is available, ie. at least the minimum available
+	// replicas required are up and running for at least minReadySeconds.
+	DeploymentAvailable DeploymentConditionType = "Available"
+	// DeploymentProgressing means the deployment is progressing. Progress for a deployment is
+	// considered when a new replica set is created or adopted, and when new pods scale up or
+	// old pods scale down. Progress is not estimated for paused deployments, when the deployment
+	// needs to rollback, or when progressDeadlineSeconds is not specified.
+	DeploymentProgressing DeploymentConditionType = "Progressing"
+	// DeploymentReplicaFailure is added in a deployment when one of its pods
+	// fails to be created or deleted.
+	DeploymentReplicaFailure DeploymentConditionType = "ReplicaFailure"
```

Comment:
Conditions are slightly different from states aka Phases: https://github.com/kubernetes/kubernetes/issues/7856

A ReplicaFailure does not mean a Deployment failure. For example. when a pod for a Deployment lands on a node that it shouldn't and it starts crashlooping, this is not a Deployment failure, is it?